File: 209-MinimumSizeSubarraySum.cpp
/*

Given an array of n positive integers and a positive integer s, find the minimal length of a contiguous subarray of which the sum ≥ s. If there isn't one, return 0 instead.

For example, given the array [2,3,1,2,4,3] and s = 7,
the subarray [4,3] has the minimal length under the problem constraint.

click to show more practice.

More practice:
If you have figured out the O(n) solution, try coding another solution of which the time complexity is O(n log n).

*/

class Solution {
public:
    int minSubArrayLen(int s, vector<int>& nums) {
        if(nums.size() == 0) return 0;
        
        int begin = 0;
        int end = 0;
        int counter = s;
        int minbegin = -1;
        int minlen = INT_MAX;
        while(end < nums.size()) {
            counter -= nums[end];
            
            while(counter <= 0) { // window found 
                if(end-begin+1 < minlen) {
                    minlen = end-begin+1; 
                    minbegin = begin;
                }
                counter += nums[begin];
                ++begin;
            }
            ++end;
        } 
        if(minbegin == -1) return 0;
        return minlen;        
    }
};

// O(nlogn) solution
class Solution1 {
public:
    int minSubArrayLen(int s, vector<int>& nums) {
        vector<int> sums = accumulate(nums);
        int n = nums.size(), minlen = INT_MAX;
        for (int i = 1; i <= n; i++) { 
            if (sums[i] >= s) {
                int p = upper_bound(sums, 0, i, sums[i] - s);
                if (p != -1) minlen = min(minlen, i - p + 1);
            }
        }
        return minlen == INT_MAX ? 0 : minlen;
    }
private:
    vector<int> accumulate(vector<int>& nums) {
        int n = nums.size();
        vector<int> sums(n + 1, 0);
        for (int i = 1; i <= n; i++) 
            sums[i] = nums[i - 1] + sums[i - 1];
        return sums;
    }
    int upper_bound(vector<int>& sums, int left, int right, int target) {
        int l = left, r = right;
        while (l < r) {
            int m = l + ((r - l) >> 1);
            if (sums[m] <= target) l = m + 1;
            else r = m;
        }
        return sums[r] > target ? r : -1;
    }
};

File: java/209-MinimumSizeSubarraySum.java
/*

Given an array of n positive integers and a positive integer s, find the minimal 
length of a contiguous subarray of which the sum ≥ s. If there isn't one, 
return 0 instead.

For example, given the array [2,3,1,2,4,3] and s = 7,
the subarray [4,3] has the minimal length under the problem constraint.

click to show more practice.

More practice:
If you have figured out the O(n) solution, try coding another solution of 
which the time complexity is O(n log n).

*/

class Solution {
    public int minSubArrayLen(int s, int[] nums) {
        if(nums==null || nums.length == 0) return 0;
        int res = Integer.MAX_VALUE;
        int begin = 0;
        int end = 0;
        int sum = 0;
        // while(begin <= end && end < nums.length) { //works too
        while(end < nums.length) {    
            sum += nums[end];
            while(sum >= s) {
                res = Math.min(res, end-begin+1);
                sum -= nums[begin];
                ++begin;
            }
            ++end;
        }
        return res == Integer.MAX_VALUE ? 0:res;
    }
}
